### PR TITLE
Telemetry level of detail

### PR DESCRIPTION
**Pull request by IvanPleshkov**
Currently, all telemetry data is too complicated. It contains a lot of data and it's difficult to read output JSON. Also, there was no data aggregation. For instance, all searches were separated into each segment and there was no aggregated searces count for all segments. This PR implements aggregated statistics and separating into different levels of detail. Descriptions of such separating can be found in development notions:
https://www.notion.so/qdrant/User-telemetry-data-2af7177a1305490a999eb9c1074b27a4
Also, there is some refactoring. Mostly for measurements and anonymizing.

### All Submissions:

* [x] Have you followed the guidelines in our Contributing document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

### New Feature Submissions:

1. [x] Does your submission pass tests?
2. [x] Have you lint your code locally using ``cargo fmt`` command prior to submission?
3. [x] Have you checked your code using ```cargo clippy``` command?

### Changes to Core Features:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your core changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

### CODE REVIEW

**Comment by agourlay** [comment]
For those interested without access to Notion, I played with it locally for a bit and here is an example of output for `detail_level=3` without anonymizing.

telemetry-level3

File: lib/collection/src/collection_manager/optimizers/indexing_optimizer.rs [hunk 1]
```diff
@@ -236,13 +236,15 @@ impl SegmentOptimizer for IndexingOptimizer {
     }
 
     fn get_telemetry_data(&self) -> OptimizerTelemetry {
-        OptimizerTelemetry::Indexing {
-            optimizations: self.get_telemetry_counter().lock().get_statistics(),
+        OptimizerTelemetry {
+            indexing: self.get_telemetry_counter().lock().get_statistics(),
+            merge: Default::default(),
```
Comment (agourlay) [inline lib/collection/src/collection_manager/optimizers/indexing_optimizer.rs:241]: are we actually tracking `merge` and `vacuum`?

File: src/common/telemetry.rs [hunk 2]
```diff
@@ -352,25 +472,84 @@ impl TelemetryCollector {
...
+        for grpc_data in &tonic_telemetry_collector.workers {
+            let lock = grpc_data.lock().get_telemetry_data();
+            result.merge(&lock);
         }
         result
     }
 }
+
+impl TelemetryData {
+    pub fn agregate(&mut self) {
```
Comment (agourlay) [inline src/common/telemetry.rs:494]:
```suggestion
    pub fn aggregate(&mut self) {
```

File: src/common/telemetry.rs [hunk 3]
```diff
@@ -24,42 +28,83 @@ pub struct TelemetryCollector {
...
-            .push(grpc_calls_aggregator.clone());
-        grpc_calls_aggregator
+    pub fn create_grpc_telemetry_collector(&mut self) -> Arc<Mutex<TonicWorkerTelemetryCollector>> {
+        let worker: Arc<Mutex<_>> = Default::default();
+        self.workers.push(worker.clone());
+        worker
     }
 }
 
 // Whole telemtry data
```
Comment (agourlay) [inline src/common/telemetry.rs:64]:
```suggestion
// Whole telemetry data
```

File: src/common/telemetry.rs [hunk 4]
```diff
@@ -352,25 +472,84 @@ impl TelemetryCollector {
...
+            result.merge(&lock);
         }
         result
     }
 }
+
+impl TelemetryData {
+    pub fn agregate(&mut self) {
+        let mut collections_short_info = CollectionShortInfoTelemetry::default();
+        for collection in self.collections.as_mut().unwrap().iter_mut() {
```
Comment (agourlay) [inline src/common/telemetry.rs:496]: Why is this unwrap safe?

**Comment by agourlay** [comment]
I did a first pass and found only minor things :+1: 

I also played with it locally, with different level of details and anonymity, it works great!
Although the 4th level of details returns a rather large payload.

I believe that we could extract important information when storing this in a document oriented like database.

File: lib/collection/src/collection_manager/optimizers/indexing_optimizer.rs (hunk 1 above)
Comment (IvanPleshkov) [reply lib/collection/src/collection_manager/optimizers/indexing_optimizer.rs:241]: We tracking all optimizers type. IMHO we can remove it from telemetry only when we remove such entities from code

File: src/common/telemetry.rs (hunk 2 above)
Comment (IvanPleshkov) [reply src/common/telemetry.rs:494]: Thanks! fixed

File: src/common/telemetry.rs (hunk 3 above)
Comment (IvanPleshkov) [reply src/common/telemetry.rs]: You have a really attentive =)

File: src/common/telemetry.rs (hunk 4 above)
Comment (IvanPleshkov) [reply src/common/telemetry.rs]: I removed unwrap usages in this function. Unwrap was safe but I believe that non-obvious "save unwraps" are bad too

**Review by generall — APPROVED** [review]

**Pull request merged**